### PR TITLE
refactor(manager/poetry): use zod schema validation

### PR DESCRIPTION
**Pull request by zeshuaro**
## Changes

Refactor to use zod schema validation for `poetry` manager

## Context

- Mainly, I want to work on the suggestion in this discussion https://github.com/renovatebot/renovate/discussions/19144, which requires some parsing of the poetry/pyproject file: 
- I've raised other PRs recently and seems like the recommendation is to use zod for validating and parsing package/lock files, so I've done some refactoring first to make changes later easier

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/poetry/__fixtures__/pyproject.2.toml [hunk 1]
```diff
@@ -5,10 +5,11 @@ description = ""
 authors = ["John Doe <john.doe@gmail.com>"]
 
 [tool.poetry.dependencies]
-dep1 = { version =  "*" }
+dep1 = { version = "*" }
```
Comment (viceice) [inline lib/modules/manager/poetry/__fixtures__/pyproject.2.toml:8]: please revert this change, it's for validation

File: lib/modules/manager/poetry/extract.ts [hunk 2]
```diff
@@ -15,34 +14,50 @@ import * as pep440Versioning from '../../versioning/pep440';
...
+  PoetryGroupRecord,
+  PoetrySchema,
+  PoetrySectionSchema,
+} from './schema';
+import { parsePoetry } from './utils';
 
 function extractFromDependenciesSection(
-  parsedFile: PoetryFile,
-  section: keyof Omit<PoetrySection, 'source' | 'group'>,
+  poetryFile: PoetrySchema,
```
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:26]: please don't rename variables unless necessary

File: lib/modules/manager/poetry/utils.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { logger } from '../../../logger';
+import { PoetrySchema, PoetrySchemaToml } from './schema';
```
Comment (viceice) [inline lib/modules/manager/poetry/utils.ts:2]:
```suggestion
import { type PoetrySchema, PoetrySchemaToml } from './schema';
```

File: lib/modules/manager/poetry/extract.ts [hunk 4]
```diff
@@ -68,35 +83,39 @@ function extractFromSection(
...
-      const path = currentValue.path;
-      const git = currentValue.git;
-      if (version) {
-        currentValue = version;
-        nestedVersion = true;
-        if (!!path || git) {
-          skipReason = path ? 'path-dependency' : 'git-dependency';
-        }
-      } else if (path) {
+      if (is.array(currentValue)) {
```
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:86]: what's the reason to change the order of this code?

Comment (zeshuaro) [reply lib/modules/manager/poetry/extract.ts:86]: We need to do this check first, otherwise we won't be able to access the properties of `currentValue` as it can be an array.

image

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/poetry/schema.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    .object({
+      requires: z.array(z.string()),
+      'build-backend': z.string().optional(),
+    })
+    .optional(),
+});
+
+export type PoetrySchema = z.infer<typeof PoetrySchema>;
+
+export const PoetrySchemaToml = Toml.pipe(PoetrySchema);
```
Comment (secustor) [inline lib/modules/manager/poetry/schema.ts:55]: Do you have considered integrating this with https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/pep621/schema.ts

That manager is already schema validating `pyproject.toml`.

File: lib/modules/manager/poetry/utils.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import { type PoetrySchema, PoetrySchemaToml } from './schema';
+
+export function parsePoetry(
+  fileName: string,
+  fileContent: string
+): PoetrySchema | null {
+  const res = PoetrySchemaToml.safeParse(fileContent);
+  if (res.success) {
+    return res.data;
+  } else {
```
Comment (secustor) [inline lib/modules/manager/poetry/utils.ts:11]: unnecessary `else`

File: lib/modules/manager/poetry/schema.ts (hunk 5 above)
Comment (zeshuaro) [reply lib/modules/manager/poetry/schema.ts:55]: Poetry doesn't follow PEP621 at the moment, so the two schemas are a bit different. So I don't think we should combine them.

**Review by viceice — APPROVED** [review]

**Comment by zharinov** [comment]
Please update the branch, as we've merged Toml schema utils separately

**Comment by zharinov** [comment]
Please let me provide my own version of this refactoring. Code changes I want to ask for are quite big, so communication could take days. So please let me try to just refactor it on my own, hopefully you'll guys like it too.

**Comment by zeshuaro** [comment]
> Please let me provide my own version of this refactoring. Code changes I want to ask for are quite big, so communication could take days. So please let me try to just refactor it on my own, hopefully you'll guys like it too.

So do I still need to do anything for this PR? Or I’ll just close it and let you finish the refactoring?

**Comment by zharinov** [comment]
Please don't, just let me try to do the refactoring myself in another PR, and then you'll be able to work on your main goal

**Comment by zharinov** [comment]
Actually, let's make it another way. Let's merge your PR and I'll make one more refactoring on top of yours.

**Review by zharinov — APPROVED** [review]

**Comment by zeshuaro** [comment]
> Actually, let's make it another way. Let's merge your PR and I'll make one more refactoring on top of yours.

Cool, I still want to work of the issue in this discussion https://github.com/renovatebot/renovate/discussions/19144

I guess I’ll just wait for you to finish the changes first then?

**Comment by zharinov** [comment]
> I guess I’ll just wait for you to finish the changes first then?

Yes, please

**Comment by zharinov** [comment]
Hi @zeshuaro, sorry for long pause. Please, take a look at the current zod-based implementation. It's not complete, but maybe 70–80% of what I wanted to achieve. Are you still interested in working on your original problem?

**Comment by zeshuaro** [comment]
No problems, I actually already had a PR raised and merged which was then reverted due to some reported issues.

There is a new PR raised to address the problem: https://github.com/renovatebot/renovate/pull/24335
